euphoria-core/src/main/java/cz/seznam/euphoria/core/executor/storage/FsSpillingListStorage.java

```diff
@@ -0,0 +1,148 @@
+  // will be spilled to disk. therefore, this list contains the lastly
+  // added elements to this list-storage
+  private final List<T> elems = new ArrayList<>();
+
+  private File storageFile;
+  private SerializerFactory.Serializer serializerInstance;
+  private SerializerFactory.Serializer.OutputStream serializerStream;
+  private boolean needsFlush;
+
+  public FsSpillingListStorage(SerializerFactory serializerFactory, int maxElemsInMemory) {
+    this.serializerFactory = Objects.requireNonNull(serializerFactory);
+    this.maxElemsInMemory = maxElemsInMemory;
+  }
+
+  @Override
+  public void add(T element) {
+    elems.add(element);
+    if (elems.size() >= maxElemsInMemory) {
+      spillElems();
+    }
+  }
+
+  private void spillElems() {
+    if (serializerStream == null) {
+      initDiskStorage();
+    }
+    for (T elem : elems) {
+      serializerStream.writeObject(elem);
+    }
+    elems.clear();
+    needsFlush = true;
+  }
+
+  @Override
+  public Iterable<T> get() {
+    return () -> {
+      if (serializerStream == null && elems.isEmpty()) {
+        return Collections.emptyIterator();
+      }
+
+      SerializerFactory.Serializer.InputStream is;
+      if (serializerStream != null) {
+        if (needsFlush) {
+          serializerStream.flush();
+          needsFlush = false;
+        }
+        try {
+          is = serializerInstance.newInputStream(new FileInputStream(storageFile));
+        } catch (FileNotFoundException e) {
+          throw new IllegalStateException("Failed to open spilling storage: " + storageFile, e);
+        }
+      } else {
+        is = null;
+      }
+
+      Iterator elemsIter = elems.iterator();
+      return new Iterator<T>() {
+        @Override
+        public boolean hasNext() {
+          boolean n = (is != null && !is.eof()) || elemsIter.hasNext();
+          if (!n && is != null) {
+            is.close();
+          }
+          return n;
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public T next() {
+          if (is != null && !is.eof()) {
+            return (T) is.readObject();
+          }
+          if (elemsIter.hasNext()) {
+            return (T) elemsIter.next();
+          }
+          if (is != null) {
+            is.close();
+          }
+          throw new NoSuchElementException();
+        }
+      };
+    };
+  }
+
+  @Override
+  public void clear() {
+    if (serializerStream != null) {
+      serializerStream.close();
+      if (!storageFile.delete()) {
+        throw new IllegalStateException("Failed to clean up storage file: " + storageFile);
+      }
+    }
+  }
+
+  private void initDiskStorage() {
+    assert storageFile == null;
+    assert serializerInstance == null;
+    assert serializerStream == null;
+    try {
+      Path workDir = FileSystems.getFileSystem(new URI("file:///")).getPath("./");
+      storageFile = Files.createTempFile(workDir, getClass().getName(), ".dat").toFile();
```

Comment:
Maybe the directory where to put this file should be configurable? If I read this correctly, it will create temp file in working directory (which could be default), but it might be useful to specify some other directory (or even several directories). If several directories would be used, then it could have performance benefits over using just a single one, if we could actually open file on each of them and then somehow (maybe completely at random) spill elements into multiple files (and therefore potentially across multiple disks).